cpp/src/io/utilities/parsing_utils.cu

```diff
@@ -0,0 +1,143 @@
+	gdf_size_type* count, ll_uint_t* positions) {
+
+	// thread IDs range per block, so also need the block id
+	const long tid = threadIdx.x + (blockDim.x * blockIdx.x);
+	const long did = tid * bytes_per_find_thread;
+	
+	const char *raw = (data + did);
+
+	const long byteToProcess = ((did + bytes_per_find_thread) < size) ?
+									bytes_per_find_thread :
+									(size - did);
+
+	// Process the data
+	for (long i = 0; i < byteToProcess; i++) {
+		if (raw[i] == key) {
+			const auto pos = atomicAdd(count, 1ull);
+			positions[pos] = did + offset + i;
+		}
+	}
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief For each of the characters in the input set, find and saves all 
+ * cccurences in the input host array of characters.
+ * The positions are stored in the output device array.
+ * 
+ * Does not load the entire file into the GPU memory at any time, so it can 
+ * be used to parse large files. Output array needs to be preallocated.
+ * 
+ * @param[in] h_data Pointer to the input character array
+ * @param[in] h_size Number of bytes in the input array
+ * @param[in] keys Vector containing the keys to count in the buffer
+ * @param[in] result_offset Offset to add to the output positions
+ * @param[out] positions Array containing the output positions
+ * 
+ * @return gdf_error with error code on failure, otherwise GDF_SUCCESS
+ *---------------------------------------------------------------------------**/
+gdf_error findAllFromSet(const char *h_data, size_t h_size, std::vector<char> keys, ll_uint_t result_offset,
+	ll_uint_t *positions) {
+
+	char* d_chunk = nullptr;
+	RMM_TRY(RMM_ALLOC (&d_chunk, min(max_chunk_bytes, h_size), 0)); 
+	
+	gdf_size_type*	d_count;
+	RMM_TRY(RMM_ALLOC((void**)&d_count, sizeof(gdf_size_type), 0) );
+	CUDA_TRY(cudaMemsetAsync(d_count, 0ull, sizeof(gdf_size_type)));
+
+	int blockSize;		// suggested thread count to use
+	int minGridSize;	// minimum block count required
+	CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&minGridSize, &blockSize, findAll) );
+
+	const size_t chunk_count = (h_size + max_chunk_bytes - 1) / max_chunk_bytes;
+	for (size_t ci = 0; ci < chunk_count; ++ci) {	
+		const auto chunk_offset = ci * max_chunk_bytes;	
+		const auto h_chunk = h_data + chunk_offset;
+		const auto chunk_bytes = std::min((size_t)(h_size - ci * max_chunk_bytes), max_chunk_bytes);
+		const auto chunk_bits = (chunk_bytes + bytes_per_find_thread - 1) / bytes_per_find_thread;
+		const int gridSize = (chunk_bits + blockSize - 1) / blockSize;
+
+		// Copy chunk to device
+		CUDA_TRY(cudaMemcpyAsync(d_chunk, h_chunk, chunk_bytes, cudaMemcpyDefault));
+
+		for (char key: keys) {
+			findAll <<< gridSize, blockSize >>> (
+				d_chunk, chunk_bytes, chunk_offset + result_offset, key,
+				d_count, positions);
+		}
+	}
+
+	RMM_TRY( RMM_FREE( d_count, 0 ) ); 
+	RMM_TRY( RMM_FREE( d_chunk, 0 ) );
+
+	CUDA_TRY( cudaGetLastError() );
+
+	return GDF_SUCCESS;
+}
```

Comment:
Missing newline at end of file. Some IDEs and tools require a line ending at the end